src/main/java/org/jabref/logic/importer/SearchBasedParserFetcher.java

```diff
@@ -39,16 +41,22 @@
             return Collections.emptyList();
         }
 
-        try (InputStream stream = getUrlDownload(getURLForQuery(query)).asInputStream()) {
+        URL urlForQuery;
+        try {
+            urlForQuery = getURLForQuery(query);
```

Comment:
Is it really necessary to extract this from the try (...) statement? I found the previous version easier to understand.

Reply:
Aim: I want to log when the URL is incorrect

![grafik](https://user-images.githubusercontent.com/1366654/91210278-40719c80-e70d-11ea-9936-dae4d3e5870c.png)

I don't know how to include it in the same try:

![grafik](https://user-images.githubusercontent.com/1366654/91210382-6bf48700-e70d-11ea-97cb-a55fecb224f5.png)

Reply:
Sorry, I cannot follow. The previous code had a `URISyntaxException` handler...

Reply:
The code before the PR had issues with code pulication. I tried to get rid of them during the PR. 

I now includ the URL in the exception with the cost of a nested try..catch

```java
        } catch (IOException e) {
            // TODO: Catch HTTP Response 401/403 errors and report that user has no rights to access resource
            try {
                throw new FetcherException("A network error occurred while fetching from " + getURLForQuery(query), e);
            } catch (URISyntaxException | MalformedURLException uriSyntaxException) {
                // does not happen
                throw new FetcherException("A network error occurred", e);
            }
        } catch (ParseException e) {
            try {
                throw new FetcherException("An internal parser error occurred while fetching from " + getURLForQuery(query), e);
            } catch (URISyntaxException | MalformedURLException uriSyntaxException) {
                // does not happen
                throw new FetcherException("An internal parser error occurred", e);
            }
        }
```

Hope, you like it! (I don't, but at least the first try stays the same)

Reply:
Well, the code duplication was introduced in the complex-query-PR. We've accepted it as a temporary solution, which should be replaced by forwarding the normal search by query method to the search by complex query method or what is probably better in the long term is to completely replace all normal text-based search fetcher by complex query fetchers. I'm not sure if your changes here really help for this.

Reply:
Note that the discussion on the code duplication is separate from the discussion how to log URLs. Sorry for doing two things in one PR - and discussing two things without making it excplicit that we are discussing two things.

In the context of URL creation and logging, I created the ADR-0014 summarizing our discussion. The full file is there: https://github.com/JabRef/jabref/blob/fix-fetchers/docs/adr/0014-separate-URL-creation-to-enable-proper-logging.md

Here are the options:

* Separate URL creation (my favorite)
* Create URL when logging the URL (IMHO your favorite)
* Include URL creation as statement before the stream creation in the try-with-resources block

Reply:
I agree if you really need the url in the catch block, then the code that you had before (creating it before the try with resource block) is the best. I'm still not convinced that you should show the URL to the user. It's only somewhat helpful for a developer fixing the fetcher tests, but not for a user (except if I overlook something). 

But that's actually a topic for a bigger discussion: where and how to log vs throw exceptions. I propose we move this discussion to the JabCon. Since if i understood you correctly, the aim of this PR is to fix some of the fetcher, I would suggest to extract all other changes (log & refactoring) to their own PR. Then we can merge this on, so that we finally release 5.1. What do you think?

Reply:
Ah, I get the point for "showing the URL". Call for an ADR ^^. https://howtodoinjava.com/best-practices/java-exception-handling-best-practices/#6

I take the freedom to choose "Separate URL creation" (and do not refactor the interface).

Reply:
Well, the merged version still contains the refactored interface, as well as the disabled tests. It would be nice if you could revist those points in one of the next evenings, when everything is a bit more quite and no coming release is hanging over our heads creating the impression of urgency.

Reply:
We included a `private` method in the interface, which reduced code clones. I will create a new issue for tracking.

Reply:
Added as issue --> https://github.com/koppor/jabref/issues/467